[onnxruntime] new ports for microsoft/onnxruntime v1.23.2

luncliff opened the pull request:
### Notes

* Resolve #36545
* Resolve #28429
* Resolve #20548

With my experience in [the registry](https://github.com/luncliff/vcpkg-registry), Some features may not work well with toolchain.

Hope more feature contributions after this work.

```cmake
# triplet example: x64-windows.cmake
message(STATUS "Using CUDA_PATH: $ENV{CUDA_PATH}")
set(TENSORRT_HOME "$ENV{CUDA_PATH}/TensorRT-10.9.0.34")
get_filename_component(TENSORRT_HOME "${TENSORRT_HOME}" ABSOLUTE)
message(STATUS "Using TENSORRT_HOME: ${TENSORRT_HOME}")
```

```console
-- Using CUDA_PATH: C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.8
-- Using TENSORRT_HOME: C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v12.8/TensorRT-10.9.0.34
```

### References

* 2024-09-06 https://github.com/microsoft/vcpkg/pull/36850#issuecomment-2334168324

#### ONNXRuntime

* https://github.com/microsoft/onnxruntime/releases/tag/v1.21.0
* https://github.com/microsoft/onnxruntime/releases/tag/v1.18.0
* https://github.com/microsoft/onnxruntime/issues/7150
* https://github.com/microsoft/onnxruntime/pull/19966
* https://github.com/microsoft/onnxruntime/pull/21492 (reverted to work with `eigen3` 3.4.0 in vcpkg)

#### microsoft/vcpkg

* `onnx`
   * #38879
* `onnxruntime-gpu`
   * #35556
   * https://github.com/microsoft/vcpkg/pull/36850#issuecomment-2223941489
   * https://github.com/microsoft/vcpkg/pull/40962
* `nvidia-cutlass`
   * https://github.com/microsoft/vcpkg/pull/39881

#### luncliff/vcpkg-registry

* https://github.com/luncliff/vcpkg-registry/pull/215
* https://github.com/luncliff/vcpkg-registry/pull/173
* https://github.com/luncliff/vcpkg-registry/pull/165
* https://github.com/luncliff/vcpkg-registry/pull/357


### Checklist

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [ ] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [ ] Only one version is in the new port's versions file.
- [ ] Only one version is added to each modified port's versions file.

dg0yt commented on ports/onnxruntime/fix-cmake.patch:
```diff
@@ -0,0 +1,140 @@
+ include(CheckCXXCompilerFlag)
+ include(CheckLanguage)
+ include(CMakeDependentOption)
+-include(FetchContent)
++include(GNUInstallDirs) # Using vcpkg. See external/onnxruntime_vcpkg_deps.cmake
+ include(CheckFunctionExists)
+ 
+ # TODO: update this once all system adapt c++20
+@@ -360,7 +360,6 @@ endif()
+ set(RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin CACHE PATH "Single output directory for all binaries.")
+ 
+ 
+-include(FetchContent)
```
I think these changes might not be needed.

luncliff replied:
Indeed. I will remove the part just after I fix the `osx | ios` build fixes.

luncliff replied:
Some targets require `include(GNUInstallDirs)` in advance. I changed the location in https://github.com/microsoft/vcpkg/pull/36850/commits/da97dd70f1de5c3df2b2b8a1196cb81790a0c846

dg0yt commented on ports/onnxruntime/fix-cmake.patch:
```diff
@@ -0,0 +1,140 @@
+       target_compile_definitions(${target_name} PRIVATE ${ORT_FLAG})
+@@ -1370,13 +1371,7 @@ endif()
+ set(ORT_BUILD_INFO "ORT Build Info: ")
+ find_package(Git)
+ if (Git_FOUND)
+-  execute_process(COMMAND ${GIT_EXECUTABLE} log -1 --format=%h
+-      WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+-      OUTPUT_VARIABLE ORT_GIT_COMMIT)
+   string(STRIP "${ORT_GIT_COMMIT}" ORT_GIT_COMMIT)
+-  execute_process(COMMAND ${GIT_EXECUTABLE} rev-parse --abbrev-ref HEAD
+-      WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+-      OUTPUT_VARIABLE ORT_GIT_BRANCH)
+   string(STRIP "${ORT_GIT_BRANCH}" ORT_GIT_BRANCH)
+   string(APPEND ORT_BUILD_INFO "git-branch=${ORT_GIT_BRANCH}, git-commit-id=${ORT_GIT_COMMIT}, ")
+ endif()
```
It might be enough to configure with `-DCMAKE_DISABLE_FIND_PACKAGE_Git=1`.

luncliff replied:
Learning from the master :D  I will that

luncliff replied:
Applied change in https://github.com/microsoft/vcpkg/pull/36850/commits/da97dd70f1de5c3df2b2b8a1196cb81790a0c846

dg0yt commented on ports/onnxruntime/portfile.cmake:
```diff
@@ -0,0 +1,158 @@
+        -Donnxruntime_USE_NEURAL_SPEED=OFF
+        -DUSE_NEURAL_SPEED=OFF
+        # for ORT_BUILD_INFO
+        -DORT_GIT_COMMIT:STRING="5f0b62cde54f59bdeac7978c9f9c12d0a4bc56db"
```
This might easily remain unchanged in an update. Maybe it could be a fake ref ( `vcpkg` or `000...`)? Or be synchronized with the actual download? (Not to mention `--head` builds.)

luncliff replied:
I would use `v${VERSION}`. I will update soon.

luncliff replied:
Applied change in https://github.com/microsoft/vcpkg/pull/36850/commits/da97dd70f1de5c3df2b2b8a1196cb81790a0c846

dg0yt commented:
IIRC the xcode generator needs a GUI session, i.e. doesn't work in CI.

luncliff commented:
Added notes for the current features.
If the PR is labeled with https://github.com/microsoft/vcpkg/labels/requires%3Aall-feature-testing,
I will change it to draft and then add more ports to the upstream.

horenmar commented on ports/onnxruntime/portfile.cmake:
```diff
@@ -0,0 +1,152 @@
+if(VCPKG_TARGET_IS_WINDOWS OR VCPKG_TARGET_IS_LINUX OR VCPKG_TARGET_IS_ANDROID)
+    vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY)
+endif()
```
From experience maintaining a private port for our company, it it possible to static link ORT on Windows and Linux, at least for a cut down build. (We did CPU inference, no training, for our prod package)

luncliff replied:
Good comment! My team always used it with DLL and a shared library.
Maybe I can remove the part and let the users customize the port output with triplet...

horenmar commented on ports/onnxruntime/portfile.cmake:
```diff
@@ -0,0 +1,152 @@
+        mpi       onnxruntime_USE_MPI
+        minimal   onnxruntime_ORT_MINIMAL_BUILD
+    INVERTED_FEATURES
+        abseil    onnxruntime_DISABLE_ABSEIL
```
I don't see some of these features in the manifest. e.g. `abseil`, or `test`.

luncliff replied:
Those are for future updates by other contributors. Honestly I couldn't test all features because the version fails to build the `test`.

luncliff commented:
Can't fix the errors of the CUDA code. I will remove `cuda` and `xnnpack` feature for this work.
I think I can support them in next `port-version`...

luncliff commented:
So the build error comes from arm_neon build.

* Reproduced in NDK 25.2.9519653
* NOT reproduced in NDK 26.2.11394342

Let me check the details...

ilya-lavrenov commented on ports/onnxruntime/fix-cmake.patch:
```diff
@@ -0,0 +1,112 @@
+     target_compile_options(onnxruntime_providers_openvino PRIVATE "-Wno-parentheses")
+   endif()
+-  add_dependencies(onnxruntime_providers_openvino onnxruntime_providers_shared ${onnxruntime_EXTERNAL_DEPENDENCIES})
++  add_dependencies(onnxruntime_providers_openvino ${onnxruntime_EXTERNAL_DEPENDENCIES})
```
could you please upstream these changes?

ilya-lavrenov replied:
Ported this patch to upstream https://github.com/microsoft/onnxruntime/pull/20053

jywu-msft replied:
i'm not following this. onnxruntime_providers_openvino should depend on onnxruntime_providers_shared. it's by design.
cc @sfatimar

luncliff replied:
The part is removed 👍

ilya-lavrenov commented on ports/onnxruntime/onnxruntime_vcpkg_deps.cmake:
```diff
@@ -0,0 +1,121 @@
+if (onnxruntime_USE_OPENVINO)
+  find_package(OpenVINO REQUIRED)
+  # deceive ENV{INTEL_OPENVINO_DIR} usages in CMakeLists.txt
+  set(ENV{INTEL_OPENVINO_DIR} "${OpenVINO_VERSION_MAJOR}.${OpenVINO_VERSION_MINOR}") # "2023.0"
```
FYI: after https://github.com/microsoft/onnxruntime/pull/19184 is merged, we don't need these WAs

luncliff replied:
Awesome! I will subscribe the PR

ilya-lavrenov commented on ports/onnxruntime/vcpkg.json:
```diff
@@ -0,0 +1,85 @@
+          "name": "openvino",
+          "default-features": false,
+          "features": [
+            "onnx"
```
you have enabled only feature to convert models from ONNX to OpenVINO IR, but all inference features like CPU, GPU are turned off, right? I suppose we need to enable at least CPU.

luncliff replied:
Added "cpu" and "gpu" in https://github.com/microsoft/vcpkg/pull/36850/commits/a3e47599d97858bda17688c52fabab59a1fe2d8a

ilya-lavrenov commented on ports/onnxruntime/vcpkg.json:
```diff
@@ -0,0 +1,85 @@
+      ]
+    },
+    "python": {
+      "description": "Enable python buildings",
```
```suggestion
      "description": "Enable python bindings",
```

luncliff commented:
> So the build error comes from arm_neon build.
> 
> * Reproduced in NDK 25.2.9519653
> * NOT reproduced in NDK 26.2.11394342
> 
> Let me check the details...

For NDK 25(Clang 14.0.7), I tried `-march=armv8.2-a+bf16` and `-march=armv8.2-a+fp16` compiler option but didn't work.
I think I haver to wait NDK 26 for Android...

* https://github.com/android/ndk/wiki/Changelog-r25
* https://releases.llvm.org/14.0.0/tools/clang/docs/LanguageExtensions.html

dg0yt commented on ports/onnxruntime/vcpkg.json:
```diff
@@ -0,0 +1,87 @@
+      "description": "Build with CoreML support",
+      "supports": "osx | ios"
+    },
+    "framework": {
+      "description": "Artifact will be Apple framework",
+      "supports": "osx | ios"
+    },
```
Should we really add such a feature? It adds much divergence from how vcpkg organizes files.

I see that qtbase has such a feature, but I wonder if it was a good idea. (And it even doesn't support ios?)

luncliff replied:
To be honest, We don't need it in vcpkg upstream

dg0yt replied:
I'm still mildly opposed to having a framework feature.

luncliff replied:
We can add it when someone wants it. I will remove it in the file

luncliff replied:
Removed in https://github.com/microsoft/vcpkg/pull/36850/commits/30ccf1fe53ab8ec4399ceeec741d41817701f389

dg0yt commented on ports/onnxruntime/portfile.cmake:
```diff
@@ -0,0 +1,149 @@
+        PACKAGES numpy sympy
+        OUT_PYTHON_VAR PYTHON3
+    )
+    get_filename_component(PYTHON_PATH "${PYTHON3}" PATH)
+else()
+    vcpkg_find_acquire_program(PYTHON3)
+    get_filename_component(PYTHON_PATH "${PYTHON3}" PATH)
+endif()
```
```suggestion
else()
    vcpkg_find_acquire_program(PYTHON3)
endif()
get_filename_component(PYTHON_PATH "${PYTHON3}" PATH)
```

dg0yt commented on ports/onnxruntime/portfile.cmake:
```diff
@@ -0,0 +1,149 @@
+        -Donnxruntime_USE_NEURAL_SPEED=OFF
+        -DUSE_NEURAL_SPEED=OFF
+        # for ORT_BUILD_INFO
+        -DORT_GIT_COMMIT:STRING="v${VERSION}"
+        -DORT_GIT_BRANCH:STRING="v${VERSION}"
```
Unless the quotes are to become part of the value:
```suggestion
        "-DORT_GIT_COMMIT:STRING=v${VERSION}"
        "-DORT_GIT_BRANCH:STRING=v${VERSION}"
```

dg0yt commented on ports/onnxruntime/portfile.cmake:
```diff
@@ -0,0 +1,149 @@
+)
+vcpkg_cmake_install()
+if(BUILD_SHARED)
+    vcpkg_cmake_config_fixup(CONFIG_PATH lib/cmake/onnxruntime PACKAGE_NAME onnxruntime)
+    vcpkg_fixup_pkgconfig() # pkg_check_modules(libonnxruntime)
```
Why is this done for static linkage?

luncliff replied:
Because the .pc files are generated in "dynamic" build

dg0yt replied:
So there are none when they are needed most...

luncliff replied:
Maybe users can customize their triplet to use it
https://github.com/microsoft/vcpkg/pull/36850#discussion_r1497523914

dg0yt commented on ports/onnxruntime/portfile.cmake:
```diff
@@ -0,0 +1,149 @@
+
+if("framework" IN_LIST FEATURES)
+    foreach(FRAMEWORK_NAME "onnxruntime.framework" "onnxruntime_objc.framework")
+        file(RENAME "${CURRENT_PACKAGES_DIR}/debug/bin/${FRAMEWORK_NAME}" "${CURRENT_PACKAGES_DIR}/debug/lib/${FRAMEWORK_NAME}")
+        file(RENAME "${CURRENT_PACKAGES_DIR}/bin/${FRAMEWORK_NAME}" "${CURRENT_PACKAGES_DIR}/lib/${FRAMEWORK_NAME}")
```
Doesn't this break CMake config?

luncliff replied:
Let me check

luncliff replied:
The part is removed in https://github.com/microsoft/vcpkg/pull/36850/commits/30ccf1fe53ab8ec4399ceeec741d41817701f389

dg0yt commented on ports/onnxruntime/portfile.cmake:
```diff
@@ -0,0 +1,149 @@
+        ORT_GIT_BRANCH
+)
+vcpkg_cmake_install()
+if(BUILD_SHARED)
```
Inconsistency: Non-standard variable here, and `VPCKG_LIBRARY_LINKAGE` in another condition further down.
```suggestion
if(VCPKG_LIBRARY_LINKAGE STREQUAL "dynamic")
```

dg0yt commented on ports/onnxruntime/onnxruntime_vcpkg_deps.cmake:
```diff
@@ -0,0 +1,121 @@
+message(STATUS "Using Dependencies from vcpkg...")
```
I guess that all `find_package` calls must be mirrored in `find_dependency` calls in the exported CMake config, at least for static library linkage.

luncliff replied:
Not sure I can add them in a simple snippet. Let me try ...

luncliff replied:
Seems like `static` build doesn't generate `onnxruntimeTargets.cmake` and `libonnxruntime.pc`.
I guess users have to deal with multiple `find_library` calls...

dg0yt replied:
That's normally fixed in vcpkg ports.

luncliff replied:
🤔 Then I should make a patch to generate `onnxruntimeTargets-*.cmake` in static build and let the file do the work.
I will test it

luncliff replied:
I overrided onnxruntimeConfig.cmake generation with the new file.
Tried to do same thing with patches, but it looked too messy (in my opinion)

luncliff commented:
* 6109c623f1e371adb64a6c497348b53ffb127328 Temporary allow failure of `onnxruntime:arm64-android`

dg0yt commented on ports/onnxruntime/portfile.cmake:
```diff
@@ -0,0 +1,150 @@
+    vcpkg_fixup_pkgconfig() # pkg_check_modules(libonnxruntime)
+endif()
+
+if(("openvino" IN_LIST FEATURES) AND VCPKG_TARGET_IS_WINDOWS)
+    file(RENAME "${CURRENT_PACKAGES_DIR}/debug/lib/onnxruntime_providers_openvino.dll" "${CURRENT_PACKAGES_DIR}/debug/bin/onnxruntime_providers_openvino.dll")
+    file(RENAME "${CURRENT_PACKAGES_DIR}/lib/onnxruntime_providers_openvino.dll" "${CURRENT_PACKAGES_DIR}/bin/onnxruntime_providers_openvino.dll")
+endif()
```
Renaming after install: Expected to break exported CMake config.

luncliff replied:
The installation was fixed in https://github.com/microsoft/onnxruntime/pull/19966
Use the latest release to apply the patch

dg0yt commented on ports/onnxruntime/project-config-template.cmake:
```diff
@@ -0,0 +1,50 @@
+# sync with portfile.cmake and onnxruntime_vcpkg_deps.cmake ...
+find_dependency(absl REQUIRED)
+find_dependency(re2 REQUIRED)
+if(@onnxruntime_BUILD_UNIT_TESTS@)
```
Generally I would recommend quotes 
```suggestion
if("@onnxruntime_BUILD_UNIT_TESTS@")
```
so that ON/OFF are never dereferenced as variables, and empty substitutions still leave valid cmake syntax.

luncliff replied:
Nice. I will apply them soon

dg0yt commented on ports/onnxruntime/portfile.cmake:
```diff
@@ -0,0 +1,150 @@
+    SOURCE_PATH "${SOURCE_PATH}/cmake"
+    OPTIONS
+        ${FEATURE_OPTIONS}
+        -DPython_EXECUTABLE:FILEPATH=${PYTHON3}
+        -DProtobuf_PROTOC_EXECUTABLE:FILEPATH=${PROTOC}
```
Quotes for paths:
```suggestion
        "-DPython_EXECUTABLE:FILEPATH=${PYTHON3}"
        "-DProtobuf_PROTOC_EXECUTABLE:FILEPATH=${PROTOC}"
```

dg0yt reviewed (commented):
A lot of work! :rocket:

luncliff commented on ports/onnxruntime/portfile.cmake:
```diff
@@ -0,0 +1,150 @@
+)
+vcpkg_cmake_install()
+vcpkg_cmake_config_fixup(CONFIG_PATH lib/cmake/onnxruntime)
+if(VCPKG_LIBRARY_LINKAGE STREQUAL "dynamic")
+    vcpkg_fixup_pkgconfig() # pkg_check_modules(libonnxruntime)
+endif()
```
.pc file template is for `libonnxruntime.so` or `libonnxruntime.dylib`. In static build, the .pc file will contain wrong library names.

* https://github.com/microsoft/onnxruntime/blob/5f0b62cde54f59bdeac7978c9f9c12d0a4bc56db/cmake/libonnxruntime.pc.cmake.in#L12
* https://github.com/microsoft/onnxruntime/blob/5f0b62cde54f59bdeac7978c9f9c12d0a4bc56db/cmake/CMakeLists.txt#L1747-L1752

dg0yt replied:
Then it could be substituted...
But I think you don't want (or need) to skip fixup. You really want to not provide the pc file for static linkage because there is no (easy) way to get the link libraries.

luncliff replied:
I can add templates and generate pc files for each targets, but I feel the port is doing too much...

I would rather wait the upstream to support the case.

dg0yt replied:
> I can add templates and generate pc files for each targets,

No...
 
(I added code to curl and gdal for generating `Libs.private` from CMake, but there the pc file had a significant relevance.)

luncliff replied:
Dropped .pc file support in https://github.com/microsoft/vcpkg/pull/36850/commits/750223d667f26baae7bf6d27da0f4b784c2ac97c

luncliff commented on ports/onnxruntime/portfile.cmake:
```diff
@@ -0,0 +1,150 @@
+if("openvino" IN_LIST FEATURES)
+    # The dependency target "onnxruntime_providers_shared" of target "onnxruntime_providers_openvino" does not exist
+    vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY)
+endif()
```
Applied in 252f11cedc74bfcc715bc141cbf8738e94a62240

dg0yt commented on ports/onnxruntime/project-config-template.cmake:
```diff
@@ -0,0 +1,50 @@
+include(CMakeFindDependencyMacro)
+
+# sync with portfile.cmake and onnxruntime_vcpkg_deps.cmake ...
+find_dependency(absl REQUIRED)
```
General in this file: Never use `REQUIRED` with `find_dependency`. It prevents this macro from doing the right thing.
```suggestion
find_dependency(absl)
```

luncliff replied:
Removed in https://github.com/microsoft/vcpkg/pull/36850/commits/750223d667f26baae7bf6d27da0f4b784c2ac97c

luncliff commented on ports/onnxruntime/portfile.cmake:
```diff
@@ -0,0 +1,150 @@
+vcpkg_copy_pdbs()
+
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
+if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
+    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/bin" "${CURRENT_PACKAGES_DIR}/bin")
+endif()
```
I guess there might some tool build based on features, but they are not defined in this port version.

luncliff commented:
Hi, requesting a review and a tag for it

vipcxj commented:
Hi I have make a working onnxruntime overlay port with cuda and tensorrt support base on your pr, I have some suggestions.
1. Instead of patch safeint everywhere, update safeint port to make it be a exported cmake package and then
```
find_package(safeint REQUIRED)
add_library(safeint_interface ALIAS safeint)
list(APPEND onnxruntime_EXTERNAL_LIBRARIES safeint_interface)
```
2. onnxruntime-cuda need cutlass 3.1.0 (not compatible with the latest 3.4.0), but cutlass 3.1.0 has a bug, it always compile the tests (spend hours), we should patch it
```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 16e0f9ae..6a242cd6 100755
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -107,7 +107,7 @@ set(CUTLASS_ENABLE_PROFILER ${CUTLASS_ENABLE_LIBRARY} CACHE BOOL "Enable CUTLASS
 set(CUTLASS_ENABLE_PERFORMANCE ${CUTLASS_ENABLE_PROFILER} CACHE BOOL "Enable CUTLASS Proformance")
 
 if(${CMAKE_PROJECT_NAME} STREQUAL ${PROJECT_NAME})
-  set(CUTLASS_ENABLE_TESTS_INIT ${CUTLASS_ENABLE_LIBRARY}})
+  set(CUTLASS_ENABLE_TESTS_INIT ${CUTLASS_ENABLE_LIBRARY})
 else()
   set(CUTLASS_ENABLE_TESTS_INIT OFF)
 endif()

```

3. onnxruntime-tensorrt need onnx2rt when onnxruntime_USE_TENSORRT_BUILTIN_PARSER is OFF (default OFF), but it need onnx2rt with branch [8.6-GA](https://github.com/onnx/onnx-tensorrt/tree/8.6-GA), not tags release/8.6-GA. 
Perhaps set onnxruntime_USE_TENSORRT_BUILTIN_PARSER to ON is an option, but I have not tried.

If you agree to change the safeint related stuff, I can submit a pr to update safeint so that it can be used directly by onnxruntime

vipcxj commented:
Currently I think the biggest difficulty in getting onnxruntime to support tensorrt is how to add tensorrt to the vcpkg system, which I solved locally by installing tensorrt globally.

luncliff commented:
Appreciate your summary, @vipcxj . I saw your work, but couldn't have a chance to review or discuss last month.

For CUDA feature(`onnxruntime[cuda,tensorrt]`) support, if we have it in future,
indeed I have troubles with latest `nvidia-cutlass` and TensorRT support in my vcpkg-registry.
That was why I cut-offed those features in this PR.
I'd like to work them in next port version, after this work is merged. (it would be `1.17.3#1` or `1.17.3#2` if I can move fast)

I agree updating `safeint` port to support `find_package(safeint REQUIRED)`. I wanted to stay in this port, but that would be more useful solution for future works.

vipcxj commented:
@luncliff I think the less patches the better, because patches can prevent port upgrades, and I even think it would be better to replace patches with regex replacements if possible.  Because once the source code is changed all the patches will be invalidated. If you don't maintain it in the future, it's not always easy to understand what you're trying to do with patches for other contributor. 
Let's take the safeint issue as an example, you changed quite a few places that use safeint, but not all of them, when I wrote my own port, I found that cuda and tensorrt part also use safeint, I searched in github, almost every sub-module uses safeint, and one page of the search results can't even show it all. This is a real burden for future upgrades.

vipcxj commented:
I found more problem at runtime.
1. In linux, vcpkg use static linkage by default. As a result, this port does not export onnxruntime_cxx_api.h directly, I have to include "onnxruntime/core/session/onnxruntime_cxx_api.h" instead of "onnxruntime_cxx_api.h"
2. There is no package onnxruntime::onnxruntime, but a lot of submodules. There are very complex dependencies between them，So linking them requires a trick, otherwise undefined link errors will occur
```cmake
    set(ONNXRUNTIME_LIBS
        onnxruntime::onnxruntime_session
        onnxruntime::onnxruntime_common
        onnxruntime::onnxruntime_flatbuffers
        onnxruntime::onnxruntime_framework
        onnxruntime::onnxruntime_graph
        onnxruntime::onnxruntime_mlas
        onnxruntime::onnxruntime_optimizer
        onnxruntime::onnxruntime_providers
        onnxruntime::onnxruntime_providers_cuda
        onnxruntime::onnxruntime_providers_shared
        onnxruntime::onnxruntime_providers_tensorrt
        onnxruntime::onnxruntime_util
    )
    target_link_libraries(${target} PRIVATE ${ONNXRUNTIME_LIBS})
    target_link_libraries(${target} PRIVATE ${ONNXRUNTIME_LIBS}) // we must link them two times.
```
3. This is the most fatal problem. A lot of Schema errors raised at runtime because the schema is registered 2 times, onnx once and onnxruntime once. onnx musted be compiled with ONNX_DISABLE_STATIC_REGISTRATION. So we can't simply add onnx as a dependency to vcpkg.json, we must download it as a submodule and compile it with ONNX_DISABLE_STATIC_REGISTRATION

dg0yt commented:
Maybe static linkage isn't very high on upstream's agenda?

> 1\. In linux, vcpkg use static linkage by default. As a result, this port does not export onnxruntime_cxx_api.h directly, I have to include "onnxruntime/core/session/onnxruntime_cxx_api.h" instead of "onnxruntime_cxx_api.h"

Is the file in a different location, or doesn't the cmake export provide the required target properties?

> 2\. There is no package onnxruntime::onnxruntime, but a lot of submodules. ...

This could be handled by a `onnxruntime::onnxruntime` interface target if upstream would accept that. 

> 3\. This is the most fatal problem. A lot of Schema errors raised at runtime because the schema is registered 2 times, onnx once and onnxruntime once. onnx musted be compiled with ONNX_DISABLE_STATIC_REGISTRATION. So we can't simply add onnx as a dependency to vcpkg.json, we must download it as a submodule and compile it with ONNX_DISABLE_STATIC_REGISTRATION

Adding another compilation of onnx probably doesn't really help: Downstream ports might still depend on both onnx and onnxruntime via (transitive) dependencies.

vipcxj commented:
@dg0yt vcpkg Is it supported to depend on multiple versions of a port? If not, onnx would be better implemented as a submodule, because with ONNX_DISABLE_STATIC_REGISTRATION and without ONNX_DISABLE_STATIC_REGISTRATION represent completely different behavior.

onnxruntime::onnxruntime exists if and only if onnxruntime_BUILD_SHARED_LIB is ON. Maybe it would be better to force onnxruntime to be dynamically linked, because the statically linked version is a bit problematic

dg0yt commented:
> Is it supported to depend on multiple versions of a port?

Dependencies may name ports and even give minimum version requirements.
In an installed tree for a given triplet, there will be exactly one version and configuration of a dependency.
But maybe your question was for a different direction.

> onnxruntime::onnxruntime exists if and only if onnxruntime_BUILD_SHARED_LIB is ON.

That is the present state...

Maybe it would be useful to add a minimal test port which builds an executable with onnxruntime, to see build-time usage issues in vcpkg CI. Example: `scripts/test_ports/vcpkg-ci-soci`.

vipcxj commented:
Hi @dg0yt
In onnxruntime_vcpkg_deps.cmake 
```cmake
if (NOT WIN32)
  find_package(unofficial-nsync CONFIG REQUIRED) # unofficial::nsync::nsync_cpp
  add_library(nsync::nsync_cpp ALIAS unofficial::nsync::nsync_cpp)
  list(APPEND onnxruntime_EXTERNAL_LIBRARIES nsync::nsync_cpp)
endif()
```
where is unofficial-nsync come from? 
And this line create a libraray nsync::nsync_cpp
``` cmake
add_library(nsync::nsync_cpp ALIAS unofficial::nsync::nsync_cpp)
```
How visible is it and can port users find it?

Since onnxruntime depends on a special onnx, which is different from onnx in vcpkg. If onnxruntime is used as a submodule, and the port user adds both onnxruntime and onnx to vcpkg.json, what happens?

dg0yt commented:
> where is unofficial-nsync come from?

It is a cmake config package for port "nsync" which only exists in vcpkg ("unofficial").

dg0yt commented:
> How visible is it and can port users find it?

User should need to care: If a port does `find_package(unofficial-foo REQUIRED)` during build, it shall also have `find_dependency(unofficial-foo)` in its exported CMake config.

vipcxj commented:
> > where is unofficial-nsync come from?
> 
> It is a cmake config package for port "nsync" which only exists in vcpkg ("unofficial").

How it is generated? If i just add nsync to vcpkg.json, I can only see nsync but not unofficial::nsync. What configuration do I need to generate this unofficial target.

dg0yt commented:
> Since onnxruntime depends on a special onnx, which is different from onnx in vcpkg. If onnxruntime is used as a submodule, and the port user adds both onnxruntime and onnx to vcpkg.json, what will happen?

If symbols are not renamed in some way, there will be two libs offering the same symbols when the user links an executable, and the linker may make different decisions then you expect.

dg0yt commented:
> How it is generated? If i just add nsync to vcpkg.json, I can only see nsync but not unofficial::nsync. What configuration do I need to generate this unofficial target.

vcpkg dependency (vcpkg port) != cmake dependency (CMake package).

The vcpkg.json is used by vcpkg (tool) to select ports to install. Ports may install CMake config files which form the CMake packages which are used by CMake build systems via `find_package`.

vipcxj commented:
> The vcpkg.json is used by vcpkg (tool) to select ports to install. Ports may install CMake config files which form the CMake packages which are used by CMake build systems via `find_package`.

I know the CMake config files installed by vcpkg generally located at share/${PACKAGE_NAME}. How does vcpkg decide whether to install share/unoffical-nsync/unoffical-nsyncConfig.cmake, nor share/nsync/nsyncConfig.cmake?

vipcxj commented:
@dg0yt So you mean since there is an unoffical-nsync, it means the port nsync just export the cmake config named unoffical-nsync, and there is no package named nsync in vcpkg system?

dg0yt commented:
> @dg0yt So you mean since there is an unoffical-nsync, it means the port nsync just export the cmake config named unoffical-nsync, and there is no package named nsync in vcpkg system?

There is a *cmake* package (named) "unoffical-nsync", provided by the *vcpkg* port (named) "nsync".

vipcxj commented:
@dg0yt Thanks

vipcxj commented on ports/onnxruntime/portfile.cmake:
```diff
@@ -0,0 +1,136 @@
+        -Donnxruntime_BUILD_SHARED_LIB=${BUILD_SHARED}
+        -Donnxruntime_BUILD_WEBASSEMBLY=OFF
+        -Donnxruntime_CROSS_COMPILING=${VCPKG_CROSSCOMPILING}
+        -Donnxruntime_USE_FULL_PROTOBUF=OFF # minimalize protoc execution
```
Currently onnxruntime_USE_FULL_PROTOBUF should always ON. Because upstream onnx only support full protobuf, and see here:
```cmake
if (NOT onnxruntime_USE_FULL_PROTOBUF)
  target_compile_definitions(onnx PUBLIC "__ONNX_NO_DOC_STRINGS")
endif()
```
The upstream onnx does not support __ONNX_NO_DOC_STRINGS.

luncliff replied:
Changed to ON in https://github.com/microsoft/vcpkg/pull/36850/commits/484fa8228219bfbceadd2283da3f2c47b97cb977

luncliff commented:
Hi, can I get additional review?

vipcxj commented:
@luncliff As I mentioned before, even if it would compile, it wouldn't work, onnx requires special compilation options, the one in vcpkg doesn't work. This is something I've verified, locally I've make a ort port with gpu and tensorrt support, but it requires quite a bit of modification to the other ports, leading to the overlay port approach being easier to use

luncliff commented:
Thanks, @vipcxj . I still have some memory after https://github.com/microsoft/vcpkg/pull/18073
But can't get the link between the macro's name `__ONNX_NO_DOC_STRINGS` and its' behavior.
Will be glad if I can fix another bug of the `onnx` port, making this PR draft until then.

It would be very helpful if I can try your test case. Is it a possible option?

vipcxj commented:
@luncliff 
The key is disable-static-registration option
```
vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
    FEATURES
        pybind11 BUILD_ONNX_PYTHON
        use-lite-proto ONNX_USE_LITE_PROTO
        disable-static-registration ONNX_DISABLE_STATIC_REGISTRATION
)
```
without it, the proto schema will be register twice, and cause many error.

Neumann-A commented:
Could it be that the real problem is thath onnxruntime tries to register the same stuff as onnx for some reason? Shouldn't the solution then that the types from onnx get removed from onnxruntime instead?

vipcxj commented:
> Could it be that the real problem is thath onnxruntime tries to register the same stuff as onnx for some reason? Shouldn't the solution then that the types from onnx get removed from onnxruntime instead?

The solution onnx guys given is use ONNX_DISABLE_STATIC_REGISTRATION to disable this behaviour on the onnx side.

BillyONeal commented:
@JavierMatosD
@vicroms
@ras0219-msft
and I discussed today.

@vicroms Would it be valid to have the suggested overlay-port for different options in a subdirectory? We don't have a way to model this today but we don't want the useful changes by @luncliff to be lost.

@ras0219-msft If it would be literally just checking in the files, they aren't going to be maintained and kept working when the library updates. Which means it would be better to just say 'go to the PR and look at what the PR did'. Because we aren't going to build / test them. On the other hand, if there are changes which can make choosing different backends easier where the majority of the relevant lines would be actually tested, that would be acceptable.

(Big conversation collapsed into) If we had infinite compute resources we would test such alternatives because that leads to 2^(number of ports doing this) number of tests, but we don't have limitless compute available.

See also: https://github.com/microsoft/vcpkg-docs/pull/332

See also: https://learn.microsoft.com/vcpkg/contributing/maintainer-guide#recommended-solutions

vipcxj commented:
@BillyONeal 
In my recent practice I have found a new problem with cuda's support for compiler versions. The latest cuda supports gcc13, but the more popular cuda only supports gcc12. One solution is to use CMAKE_CUDA_HOST_COMPILER to individually specify the compiler version of the cuda. But for now the user specified CMAKE_CUDA_HOST_COMPILER is not passed into the port. i tried customizing the triplet and specifying it in the triplet. But still it doesn't work. I ended up modifying the port's vcpkg_cmake_configure with -DCMAKE_CUDA_HOST_COMPILER=${CMAKE_CUDA_HOST_COMPILER} to finally fix the problem. This means that modifying the portfile is unavoidable in order to pass CMAKE_CUDA_HOST_COMPILER to the port. Is there a better way to do this?

luncliff commented:
Sorry for no update since last comment. I'm working with the later version, ONNXRuntime 1.18.0. Will update here as soon as the work becomes ready.

* https://github.com/microsoft/onnxruntime/releases/tag/v1.18.0

BillyONeal commented:
Probably conflicts with `onnxruntime-gpu`. See also https://github.com/microsoft/vcpkg/pull/38275 and https://github.com/microsoft/vcpkg/pull/38879

WangWeiLin-MV commented on ports/onnxruntime/onnxruntime_vcpkg_deps.cmake:
Thank you for your numerous contributions. Compared with `onnxruntime_external_deps.cmake`, the dependency provided by vcpkg is changed here. This could be integrated into upstream: directly replaced, or an option like `DEPS_BY_VCPKG` to switch `onnxruntime_vcpkg_deps.cmake`, or any other methods. I'm not sure which method is best, but it would be better to reuse the test components in the upstream repository.

The same is `fix-cmake.patch` and `fix-sources.patch`.

WangWeiLin-MV reviewed (dismissed):
The ports `onnx`, `onnxruntime`, `onnx-optimizer`, `libtorch`, `openvino` usage test pass with the following triplets under `nvcc` 12.5 and `cuDNN` 9.2.0:

* x64-windows

JavierMatosD commented:
It appears that the `onnxruntime` port is conflicting with the `onnxruntime-gpu` port. As it stands, we can only have one of these ports in our registry. 


### Files Produced by Both `onnxruntime` and `onnxruntime-gpu`:

1. **Binaries**:
   - /bin/onnxruntime.dll
   - /bin/onnxruntime.pdb
   - /bin/onnxruntime_providers_shared.dll
   - /bin/onnxruntime_providers_shared.pdb
   - /debug/bin/onnxruntime.dll
   - /debug/bin/onnxruntime.pdb
   - /debug/bin/onnxruntime_providers_shared.dll
   - /debug/bin/onnxruntime_providers_shared.pdb

2. **Libraries**:
   - /debug/lib/onnxruntime.lib
   - /debug/lib/onnxruntime_providers_shared.lib
   - /lib/onnxruntime.lib
   - /lib/onnxruntime_providers_shared.lib

3. **Include Headers**:
   - /include/onnxruntime/cpu_provider_factory.h
   - /include/onnxruntime/onnxruntime_c_api.h
   - /include/onnxruntime/onnxruntime_cxx_api.h
   - /include/onnxruntime/onnxruntime_cxx_inline.h
   - /include/onnxruntime/onnxruntime_float16.h
   - /include/onnxruntime/onnxruntime_run_options_config_keys.h
   - /include/onnxruntime/onnxruntime_session_options_config_keys.h

4. **Shared Files**:
   - /share/onnxruntime/copyright
   - /share/onnxruntime/vcpkg.spdx.json
   - /share/onnxruntime/vcpkg_abi_info.txt

### Files Produced Only by `onnxruntime-gpu`:

1. **Additional Binaries**:
   - /bin/onnxruntime_providers_cuda.dll
   - /bin/onnxruntime_providers_cuda.pdb
   - /bin/onnxruntime_providers_tensorrt.dll
   - /bin/onnxruntime_providers_tensorrt.pdb
   - /debug/bin/onnxruntime_providers_cuda.dll
   - /debug/bin/onnxruntime_providers_cuda.pdb
   - /debug/bin/onnxruntime_providers_tensorrt.dll
   - /debug/bin/onnxruntime_providers_tensorrt.pdb

2. **Additional Libraries**:
   - /debug/lib/onnxruntime_providers_cuda.lib
   - /debug/lib/onnxruntime_providers_tensorrt.lib
   - /lib/onnxruntime_providers_cuda.lib
   - /lib/onnxruntime_providers_tensorrt.lib

3. **Additional Include Headers**:
   - /include/experimental_onnxruntime_cxx_api.h
   - /include/experimental_onnxruntime_cxx_inline.h
   - /include/onnxruntime_lite_custom_op.h
   - /include/onnxruntime_training_c_api.h
   - /include/onnxruntime_training_cxx_api.h
   - /include/onnxruntime_training_cxx_inline.h
   - /include/provider_options.h
   - /include/tensorrt_provider_factory.h
   
Are the exported symbols in the `onnxruntime.dll` and `onnxruntime.lib` files different between the release version [(used in `onnxruntime-gpu`)](https://github.com/microsoft/vcpkg/blob/adb391bd2cb05fdaef1ccef375dd50d998f01cc0/ports/onnxruntime-gpu/portfile.cmake#L3C1-L7C2) and the version in `onnxruntime` port?

If the exported symbols are the same, we may be able to merge the extra functionality from `onnxruntime-gpu` into the `onnxruntime` port via features.

luncliff commented:
Hi team. I was waiting the upstream's pull request, but seems like it's taking more time than I expected.  
I just squashed changes into several commits and rebased with the `master` branch again.

Here is the current status of this pull request. (In my memory, feel free to comment more!)

### Notes

* ONNX 1.16.0+ CMake build uses `ONNX` namespace. I made the other ports to use it.
  * https://github.com/onnx/onnx/blob/990217f043af7222348ca8f0301e17fa7b841781/CMakeLists.txt#L714-L730
* I suggested ONNXRuntime upstream to support `onnxruntime_USE_VCPKG` so we can inject existing ports.
  * https://github.com/microsoft/onnxruntime/pull/21348
  * One of the concerns is `eigen3` 3.4.1 update.  
    See https://github.com/microsoft/onnxruntime/blob/605a84ffc96efb50c1d9b45ec6e62fef947f6c45/cmake/deps.txt#L20-L25

### Issues

* `onnx` should be built with `ONNX_DISABLE_STATIC_REGISTRATION` CMake option to work properly.
  * https://github.com/microsoft/vcpkg/pull/36850#issuecomment-2099781647
  * https://github.com/microsoft/vcpkg/pull/36850#issuecomment-2122110102
  * https://github.com/microsoft/vcpkg/pull/38879

As a vcpkg user and registry maintainer, I think https://github.com/microsoft/vcpkg/pull/38879 can be closed.  
We can add warning message in `onnxruntime` portfile so the user can be more cautious. And it will be better if we can reduce burdun of port maintenance by doing so.


* `onnxruntime-gpu` installs CUDA and TensorRT DLLs. (`onnxruntime_providers_cuda` & `onnxruntime_providers_tensorrt`)
  * https://github.com/microsoft/vcpkg/pull/36850#issuecomment-2223941489
  * https://dev.azure.com/vcpkg/public/_build/results?buildId=106399&view=results

Building CUDA in vcpkg CI was simple. However, installation path of TensorRT is unknown.  
I made the portfile.cmake to use variable `TENSORRT_ROOT`, forwarding it to `onnxruntime_TENSORRT_HOME` CMake option.  
(Maybe user can declare it in the triplet)

luncliff commented on ports/onnxruntime/portfile.cmake:
```diff
@@ -0,0 +1,172 @@
+file(COPY "${CMAKE_CURRENT_LIST_DIR}/onnxruntime_external_deps.cmake" DESTINATION "${SOURCE_PATH}/cmake/external")
+file(COPY "${CMAKE_CURRENT_LIST_DIR}/project-config-template.cmake" DESTINATION "${SOURCE_PATH}/cmake")
+
+# https://github.com/microsoft/onnxruntime/blob/26250ae74d2c9a3c6860625ba4a147ddfb936907/cmake/deps.txt#L20-L25
+vcpkg_from_gitlab(
+    GITLAB_URL https://gitlab.com
+    OUT_SOURCE_PATH EIGEN_SOURCE_PATH
+    REPO libeigen/eigen
+    REF e7248b26a1ed53fa030c5c459f7ea095dfd276ac
+    SHA512 2ede6fa56b8374cd5618a9cca9f3666909255277d0fe23eb54266972a9ab4e6c8c00abcb8fab918ea45b4aec37a3d316ca01107ff082dc9f19851df58d7ac80d
+)
```
Here, we are using a different version of `eigen3`.

WangWeiLin-MV replied:
This will introduce potential dependency conflicts and https://learn.microsoft.com/en-us/vcpkg/contributing/maintainer-guide#do-not-use-vendored-dependencies

Could the required fixes be backported?

WangWeiLin-MV replied:
Or temporarily change the port `eigen3` to the commit first, and then wait for the upstream release.

luncliff replied:
I don't think I can backport the related part correctly in ONNXRuntime. Either bumping `eigen3` for the other ports.

In my vcpkg-registry, I can move to any commit I want, but it's different in this repository.
So I need some direction for the next move.

* To bump `eigen3`, it should be done in another PR considering the work size
* To backport `eigen3`, I can try, but some API might block coupled to the version and I can't make a correct change in the case

WangWeiLin-MV replied:
I would like the simple upgrade but upstream hasn’t released new version. You could keep this usage temporarily while fixing onnx.

If `eigen3` is still not published before merging current PR, I will need further review by the team.

BillyONeal replied:
I don't think vendoring a different eigen than the one selected by vcpkg is acceptable.

BillyONeal replied:
Based on the comment "# This Eigen commit id matches the eigen archive being consumed from https://gitlab.com/libeigen/eigen/-/archive/3.4/eigen-3.4.zip" given that vcpkg is consuming that same thing the one in our port should have the fixes that onnxruntime needs?

Do we know of a specific change onnxruntime needs that we could backport to our port?

dg0yt commented:
> I suggested ONNXRuntime upstream to support onnxruntime_USE_VCPKG so we can inject existing ports.

I prefer to recommend upstreams to not use vcpkg-specific hacks. There is nothing special about wanting de-vendored dependencies. (vcpkg may have unofficial CMake configs, but they are unofficial to limit their proliferation.)

luncliff commented:
> I prefer to recommend upstreams to not use vcpkg-specific hacks. There is nothing special about wanting de-vendored dependencies. (vcpkg may have unofficial CMake configs, but they are unofficial to limit their proliferation.)

Most of `FetchContent` and `find_pacakge` are integrated in upstream.
Currently, there are a few target mismatches between the source build and the vcpkg installation.
When those changes are gone, `onnxruntime_USE_VCPKG ` can be removed and the required changes will be spread into each CMake modules in the project.

I think it's enough for now.

BillyONeal commented on ports/onnxruntime-gpu/portfile.cmake:
```diff
@@ -1,102 +1,2 @@
-# # Handle copyright
-vcpkg_install_copyright(FILE_LIST "${SOURCE_PATH}/onnxruntime-win-x64-gpu-${VERSION}/LICENSE")
+set(VCPKG_POLICY_EMPTY_PACKAGE enabled)
+message(WARNING "${PORT} is deprecated. Please use port onnxruntime instead.")
```
Should we just outright deindex it?

luncliff replied:
Any rules for deindexing? I simply followed the case of `qt`

BillyONeal replied:
As written is OK; it's more a user experience thing. If you think there are lots of existing users of `onnxruntime-gpu` who would be confused by it suddenly not existing, then keeping this makes sense. That was the case with qt, in particular because `vcpkg install qt` is something one is likely to do.

If you think there aren't a huge number of users who would be confused by it not existing, making it not exist might be clearer for most users because they won't be shown that a port exists that they shouldn't use.

To make it not exist:

* Revert all changes to `versions/o-/onnxruntime-gpu.json`
* Delete the `onnxruntime-gpu` entry from `versions/baseline.json`
* Delete `ports/onnxruntime-gpu`

Otherwise it's fine as written

BillyONeal commented:
@luncliff Do you want us to explicitly not take https://github.com/microsoft/vcpkg/pull/40962 ? I agree that you're fixing more stuff in here but it isn't clear to me that blocking this 'simple version bump' is the right thing pending your other fixes here...   Thanks!

luncliff commented:
I'm fine with #40962 and other updates of related ports. I'm checking them regularly.

I will try supporting the current `eigen3` in vcpkg. Then I will move to ONNX Runtime 1.19.2.

BillyONeal commented:
> I'm fine with #40962

Ok, I merged it then. Thanks!

luncliff commented:
Status update:

Another blocker appeared in https://github.com/microsoft/vcpkg/pull/36850/commits/be8880ae38eecf975539610fa904bac1c88dfa90

* In ONNXRuntime 1.19.2, `onnxruntime[cuda]` require later version of `nvidia-cutlass`
  * However, to use the later version of `nvidia-cutlass`, `libtorch` need to be updated.  
    https://github.com/microsoft/vcpkg/pull/39881#issuecomment-2226967262

Changing `libtorch` can be big work to put in this PR. I'm considering ci.baseline.txt change.

BillyONeal commented:
/azp run

BillyONeal commented:
Rerunning due to network mistake last night.

vipcxj commented:
I've updated vcpkg in the last couple of days, and as a result the onnxruntime overlay, which had been working before, is no longer compiling. After looking for the cause so far, I found two:
1. The export name of the port nsync has been changed.
2. protobuf's port has been upgraded, and there is a break change, causing onnxruntime to fail to compile, and the related [issue](https://github.com/microsoft/onnxruntime/issues/21308) is still open.

luncliff commented:
Ah, my bad... I forgot to change this into draft mode.
I was suffering with the `libtorch` and its dependencies because CUTLASS was a blocker for CUDA build.

I will check the port updates and rebase again...

vipcxj commented:
I found a new compile error, `no member named 'memcpy' in namespace 'std' `. The solution is simple, just add the missing header file `<cstring>`. What I'm curious about is that this error wasn't there before, and I didn't change the version of onnxruntime, it's still 1.17.3. I looked at the most recent source code, and none of the tags fixed the problem, but the latest main branch did. I've locally patched a patch for version 1.17.3, hope that helps you!
```
diff --git a/onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc b/onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc
index e8a5855b36..3e3fb7eb88 100644
--- a/onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc
+++ b/onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc
@@ -5,5 +5,6 @@

 #include <algorithm>
 #include <cassert>
+#include <cstring>
 #include <iostream>
 #include <unordered_map>

```

luncliff committed:
[onnxruntime] create a new port with 1.21.0

* https://github.com/microsoft/onnxruntime/releases/tag/v1.21.0

luncliff committed:
[onnxruntie-gpu] redirect to onnxruntime[cuda,tensorrt]

luncliff committed:
[onnxruntime] update to 1.22.1

- use cudnn-frontend port
- vcpkg_cmake_config_fixup only if linkage is dynamic
- Eigen3::Eigen for include_directories(BEFORE)
- disable uwp, emscripten
- https://github.com/microsoft/onnxruntime/releases/tag/v1.22.1
- https://github.com/luncliff/vcpkg-registry/pull/426

luncliff commented:
Disabled UWP, Emscripten build

The port `onnxruntime-gpu` installs CUDA and TensorRT provide DLLs. For `onnxruntime[tensorrt]`, The portfile.cmake needs a path to the installed location...

JoergAtGithub commented:
@luncliff Thanks for the great work!

There is the following error that let x86_windows CI fail:
```
D:\a\_work\1\s\ports\onnxruntime\vcpkg.json: error: the "version>=" constraint to onnx names version 1.16.0 which does not exist in the version database. All versions must exist in the version database to be interpreted by vcpkg.
D:\a\_work\1\s\versions\o-\onnx.json: note: consider removing the version constraint or choosing a value declared here
```

AFAIK ports that are expected to fail on CI, needs to be listed in https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt

dg0yt commented:
The problem is that a *literal* `1.16.0` doesn't exist. But this is required to determine the version-scheme. If you have to use version constraints, use `1.16.2`.

luncliff committed:
[onnxruntime] change to 1.23.2, use ONNX 1.19.0

luncliff committed:
[onnxruntime] fix configure error with abseil-cpp 20250814

luncliff committed:
[onnxruntime] disable CUDA in static linkage

JoergAtGithub commented on scripts/ci.feature.baseline.txt:
I guess you just need to delete this 2 entries to make the UWP builds pass.

JoergAtGithub commented:
> I got this error on Linux:
> 
> ```
> FAILED: [code=1] CMakeFiles/onnxruntime_common.dir/home/jenkins/workspace/ext-libs/src/vcpkg/buildtrees/onnxruntime/src/v1.23.2-3e8400852c.clean/onnxruntime/core/common/spin_pause.cc.o 
> /usr/lib/llvm-21/bin/clang++ -DCPUINFO_SUPPORTED_PLATFORM=1 -DEIGEN_MPL2_ONLY -DEIGEN_USE_THREADS -DENABLE_CPU_FP16_TRAINING_OPS -DHAS_STRING_VIEW=1 -DONLY_C_LOCALE=0 -DORT_ENABLE_STREAM -DPLATFORM_POSIX -D_GNU_SOURCE -I/home/jenkins/workspace/ext-libs/src/vcpkg/buildtrees/onnxruntime/src/v1.23.2-3e8400852c.clean/include/onnxruntime -I/home/jenkins/workspace/ext-libs/src/vcpkg/buildtrees/onnxruntime/src/v1.23.2-3e8400852c.clean/include/onnxruntime/core/session -I/home/jenkins/workspace/ext-libs/src/vcpkg/buildtrees/onnxruntime/x64-linux-bitwig-rel -I/home/jenkins/workspace/ext-libs/src/vcpkg/buildtrees/onnxruntime/src/v1.23.2-3e8400852c.clean/onnxruntime -I/home/jenkins/workspace/ext-libs/src/vcpkg/installed/x64-linux-bitwig/include/eigen3 -isystem /home/jenkins/workspace/ext-libs/src/vcpkg/installed/x64-linux-bitwig/include -fPIC -pipe -mavx2 -mfma -mf16c -ffunction-sections -fdata-sections -DCPUINFO_SUPPORTED -DNDEBUG -O3  -O3 -DNDEBUG -fPIC -Wall -Wextra -Wno-unused-parameter -Wno-deprecated-copy -Wno-tautological-pointer-compare -Wno-ambiguous-reversed-operator -Wno-deprecated-anon-enum-enum-conversion -Wno-deprecated-enum-enum-conversion -Wno-undefined-var-template -Wno-deprecated-builtins -Wno-deprecated-literal-operator -Wshorten-64-to-32 -mwaitpkg -MD -MT CMakeFiles/onnxruntime_common.dir/home/jenkins/workspace/ext-libs/src/vcpkg/buildtrees/onnxruntime/src/v1.23.2-3e8400852c.clean/onnxruntime/core/common/spin_pause.cc.o -MF CMakeFiles/onnxruntime_common.dir/home/jenkins/workspace/ext-libs/src/vcpkg/buildtrees/onnxruntime/src/v1.23.2-3e8400852c.clean/onnxruntime/core/common/spin_pause.cc.o.d -o CMakeFiles/onnxruntime_common.dir/home/jenkins/workspace/ext-libs/src/vcpkg/buildtrees/onnxruntime/src/v1.23.2-3e8400852c.clean/onnxruntime/core/common/spin_pause.cc.o -c /home/jenkins/workspace/ext-libs/src/vcpkg/buildtrees/onnxruntime/src/v1.23.2-3e8400852c.clean/onnxruntime/core/common/spin_pause.cc
> /home/jenkins/workspace/ext-libs/src/vcpkg/buildtrees/onnxruntime/src/v1.23.2-3e8400852c.clean/onnxruntime/core/common/spin_pause.cc:38:68: error: too few arguments to function call, expected 3, have 2
>    38 |     __builtin_ia32_tpause(0x0, __rdtsc() + tpause_spin_delay_cycles);
>       |     ~~~~~~~~~~~~~~~~~~~~~                                          ^
> 1 error generated.
> ```
> 
> Note: this doesn't happen with gcc.

Upstream onnxruntime lists Linux/Clang as "Not tested":
<img width="773" height="193" alt="grafik" src="https://github.com/user-attachments/assets/684610cf-7e56-408e-ae75-f7b392778380" />
see: https://onnxruntime.ai/docs/build/inferencing.html#supported-architectures-and-build-environments

luncliff committed:
[onnxruntime] support wasm32-emscripten build

luncliff committed:
[onnxruntime-gpu] sync with onnxruntime port

luncliff committed:
ci: feature-fails for onnxruntime[tensorrt]

luncliff commented:
Hi, vcpkg team. Would you help me using ci.feature.baseline.txt file for `onnxruntime[tensorrt]`?
I guess the vcpkg CI environment doesn't have the NVIDIA TensorRT installed, so the port feature test should fail with configuration error. `onnxruntime-gpu` requires 2 features(`cuda` and `tensorrt`), therefore it will cascade.

BillyONeal commented:
> the traditional full CI part still includes the failing feature set (due to default features or to dependencies)

Specifically, `onnxruntime-gpu` depends on `onnxruntime[tensorrt]` which causes that feature to be turned on. It would need to be `=skip`'d

luncliff committed:
ci: onnxruntime-gpu = skip

luncliff committed:
ci: onnxruntime fails in ci.baseline.txt

luncliff committed:
Revert "ci: onnxruntime fails in ci.baseline.txt"

This reverts commit c2234d852eb7bcae1d04689a47a91327a86ed32f.

luncliff committed:
ci: onnxruntime fails in ci.baseline.txt

BillyONeal committed:
Merge remote-tracking branch 'origin/master' into HEAD

BillyONeal committed:
Format

BillyONeal committed:
Merge remote-tracking branch 'origin/master' into HEAD

luncliff commented:
Sorry for taking this long. There was a series of suffering in my life after opening this PR. Thankfully it didn't crush me.
I will monitor the build failure issues for a while.

patrikhuber commented:
I've just tried to use the `onnxruntime` port, I'm on the latest vcpkg 12.12.2025 baseline.

When trying to load an onnx model, it throws an exception and prints a few hundred warnings like these onto the terminal:

```
Schema error: Trying to register schema with name Min (domain:  version: 8) from file C:\Users\Patrik\Projects\myproject\out\build\x64-debug\vcpkg_installed\vcpkg\blds\onnx\src\v1.19.0-c251c1eccc.clean\onnx\defs\math\old.cc line 3909, but it is already registered from file C:\Users\Patrik\Projects\myproject\out\build\x64-debug\vcpkg_installed\vcpkg\blds\onnx\src\v1.19.0-c251c1eccc.clean\onnx\defs\math\old.cc line 3909

Schema error: Trying to register schema with name Max (domain:  version: 8) from file C:\Users\Patrik\Projects\myproject\out\build\x64-debug\vcpkg_installed\vcpkg\blds\onnx\src\v1.19.0-c251c1eccc.clean\onnx\defs\math\old.cc line 3907, but it is already registered from file C:\Users\Patrik\Projects\myproject\out\build\x64-debug\vcpkg_installed\vcpkg\blds\onnx\src\v1.19.0-c251c1eccc.clean\onnx\defs\math\old.cc line 3907

Schema error: Trying to register schema with name Sum (domain:  version: 8) from file C:\Users\Patrik\Projects\myproject\out\build\x64-debug\vcpkg_installed\vcpkg\blds\onnx\src\v1.19.0-c251c1eccc.clean\onnx\defs\math\old.cc line 1588, but it is already registered from file C:\Users\Patrik\Projects\myproject\out\build\x64-debug\vcpkg_installed\vcpkg\blds\onnx\src\v1.19.0-c251c1eccc.clean\onnx\defs\math\old.cc line 1588

Schema error: Trying to register schema with name Mean (domain:  version: 8) from file C:\Users\Patrik\Projects\myproject\out\build\x64-debug\vcpkg_installed\vcpkg\blds\onnx\src\v1.19.0-c251c1eccc.clean\onnx\defs\math\old.cc line 1598, but it is already registered from file C:\Users\Patrik\Projects\myproject\out\build\x64-debug\vcpkg_installed\vcpkg\blds\onnx\src\v1.19.0-c251c1eccc.clean\onnx\defs\math\old.cc line 1598

Schema error: Trying to register schema with name MaxPool (domain:  version: 8) from file C:\Users\Patrik\Projects\myproject\out\build\x64-debug\vcpkg_installed\vcpkg\blds\onnx\src\v1.19.0-c251c1eccc.clean\onnx\defs\nn\old.cc line 2492, but it is already registered from file C:\Users\Patrik\Projects\myproject\out\build\x64-debug\vcpkg_installed\vcpkg\blds\onnx\src\v1.19.0-c251c1eccc.clean\onnx\defs\nn\old.cc line 2492
```

A few remarks:

1) The terminal output shows v1.19.0: `vcpkg_installed\vcpkg\blds\onnx\src\v1.19.0-c251c1eccc.clean`, but vcpkg clearly installed 1.23.2:
```
1> [CMake]     onnxruntime:x64-windows@1.23.2 -- git+https://github.com/microsoft/vcpkg@7391737791bbf21b648ab4e95521f31fdf502d59
```

2) While building onnxruntime, it says:
<img width="1379" height="175" alt="image" src="https://github.com/user-attachments/assets/8e534d9c-dce6-4a61-bb8d-047d93a826e3" />
It seems that that could be the root of the issue - this flag could be missing when building the "onnx" package in vcpkg, which then results in those "already registered" errors?

But if you read the error closely, it actually lists the same filename twice, so if one is to believe the error message, the same file has registered the schema twice?

3) I'm not clear whether I should use the onnxruntime-gpu package instead and how they differ. I am guessing onnxruntime-gpu might be deprecated in favour of onnxruntime, but I am not certain.

Any help, am I missing something?

JoergAtGithub commented:
These are two different packages:

- ONNX is v1.19.0
- ONNX-Runtime is v1.23.2

Don't use onnxruntime-gpu, this is just a wrapper for backward compatibility.

vipcxj commented:
@patrikhuber I encountered this issue a long time ago; please see my comment
https://github.com/microsoft/vcpkg/issues/36850#issuecomment-2099781647_
